### PR TITLE
[netatmo] Migration of Netatmo to bnd

### PR DESCRIPTION
**Pull request by clinique**
Signed-off-by: Gaël L'hopital <gael@lhopital.org>

### CODE REVIEW

**Comment by davidgraeff** [comment]
Looks like the old .project file did not get removed

**Comment by lolodomo** [comment]
You need to reformat the binding POM file.

**Comment by clinique** [comment]
@lolodomo : I took the opportunity of this migration to remove to dependencies on local jars : commons-codec and google gson, switching to platform provided versions.

File: bundles/org.openhab.binding.netatmo/.classpath [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+		<attributes>
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="con" path="org.eclipse.m2e.MAVEN2_CLASSPATH_CONTAINER">
+		<attributes>
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="lib" path="lib/tchaikovsky.jar" sourcepath="lib/tchaikovsky-sources.jar" />
```
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/.classpath]: Be careful, you referenced a wrong library. You need an entry for the library netatmo-java-retrofit-1.1.3.jar

**Comment by clinique** [comment]
> .classpath must be fixed.
> 
> Please fix the indentation in the file bundles/org.openhab.binding.netatmo/pom.xml
> 
> There were previously 9 libraries in lib. One is remaining and you added 6 maven dependencies. So what became to the 9-6-1=2 other libraries ?

I took the opportunity of this migration to remove to dependencies on local jars : commons-codec and google gson, switching to platform provided versions.

File: features/karaf/openhab-addons/src/main/feature/feature.xml [hunk 2]
```diff
@@ -513,7 +513,13 @@
...
+        <bundle dependency="true">wrap:mvn:com.squareup.okhttp/okhttp/2.3.0$Bundle-Name=OkHttp&amp;Bundle-SymbolicName=com.squareup.okhttp&amp;Bundle-Version=2.3.0</bundle>
+        <bundle dependency="true">wrap:mvn:com.squareup.okio/okio/1.3.0$Bundle-Name=Okio&amp;Bundle-SymbolicName=com.squareup.okio&amp;Bundle-Version=1.3.0</bundle>
+        <bundle dependency="true">wrap:mvn:com.squareup.retrofit/retrofit/1.9.0$Bundle-Name=Retrofit&amp;Bundle-SymbolicName=com.squareup.retrofit&amp;Bundle-Version=1.9.0</bundle>
+        <bundle dependency="true">wrap:mvn:org.apache.oltu.oauth2/org.apache.oltu.oauth2.client/1.0.0$Bundle-Name=Apache%20Oltu%20-%20OAuth%202.0%20-%20Client&amp;Bundle-SymbolicName=oorg.apache.oltu.oauth2-client&amp;Bundle-Version=1.0.0</bundle>
+        <bundle dependency="true">wrap:mvn:org.apache.oltu.oauth2/org.apache.oltu.oauth2.common/1.0.0$Bundle-Name=Apache%20Oltu%20-%20OAuth%202.0%20-%20Common&amp;Bundle-SymbolicName=org.apache.oltu.oauth2-common&amp;Bundle-Version=1.0.0</bundle>
+        <bundle start-level="80">mvn:org.openhab.addons.bundles/org.openhab.binding.netatmo/${project.version}</bundle>
```
Comment (J-N-K) [inline features/karaf/openhab-addons/src/main/feature/feature.xml]: I think both oauth bundles are already OSGI bundles. Wrapping is not needed then. Could you check the other bündle, too? You can download the .jar, rename it to .zip and then check the MANIFEST.MF in META-INF. If it has `Bundle-Manifest-Version: 2` in it, you don't need to wrap it.

**Comment by lolodomo** [comment]
In classpathentry you put path and sourcepath. In Rfxcom binding, I only used path. What is correct  ?

**Comment by J-N-K** [comment]
If you include the jar and the sources.jar you can add the sources.jar to sourcePath. Then Eclipse can show the source of your included library. For resolving the path is enough.

File: features/karaf/openhab-addons/src/main/feature/feature.xml (hunk 2 above)
Comment (clinique) [reply features/karaf/openhab-addons/src/main/feature/feature.xml]: Seems good, changing features accordingly

**Comment by lolodomo** [comment]
That looks good now, doesn't it ?
@clinique: did you test the binding after migration?

**Comment by clinique** [comment]
Well, it looked so. Testing it in running environment, I've got an issue with com.squareup.okhttp : 
`
2019-04-03 10:52:29.332 [WARN ] [org.apache.felix.fileinstall        ] - Error while starting bundle: file:/usr/share/openhab2/addons/org.openhab.binding.netatmo-2.5.0-SNAPSHOT.jar
org.osgi.framework.BundleException: Could not resolve module: org.openhab.binding.netatmo [273]
  Unresolved requirement: Import-Package: com.squareup.okhttp

	at org.eclipse.osgi.container.Module.start(Module.java:444) ~[?:?]
	at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:383) ~[?:?]
`

@davidgraeff : I've tried to find the origin of this without success. Would you have a hint ?

**Comment by davidgraeff** [comment]
Parts of the dependencies especially okhttp are not OSGi ready bundles.

You find out if a bundle is an OSGi bundle, by searching with google for "com.squareup.okhttp" and going to maven central or jcenter. Download the offered .jar file for the bundle and open it with a zip archiver. Look for META-INF/manifest.mf and if there is any "export" line in it.

All bundles that are not OSGi need to be wrapped. Add "wrap:" infront of the "mvn:" address. See other examples in the `features/karaf/.../features.xml` file.

**Comment by davidgraeff** [comment]
Can you use google gson, I mean the real one? Because that is part of the openhab base dependencies already. At the moment you reference org.json from 2013?!

**Comment by clinique** [comment]
When something works, don't change it...
I would prefer finalize the migration before making any further changes, while I agree in the principle.

**Comment by davidgraeff** [comment]
Can you fix the conflicts please

**Comment by J-N-K** [comment]
Restarted Travis. The build was green though,seems to be a report problem.

**Comment by J-N-K** [comment]
Can someone confirm that the last version is working?

**Comment by clinique** [comment]
> Can someone confirm that the last version is working?

After testing I confirm that this merged version does not work as expected. I face issues with okhttp dependency, and then android.net ...
@lolodomo FYI, I'm switching back to pre bnd migration in order to test an alternative to okhttp libs.

**Comment by J-N-K** [comment]
I didn't notice the missing <scope> in your dependencies. Please try to add `<scope>provided</scope>`, recompile and test that jar.

**Comment by J-N-K** [comment]
I‘ll have a look later this evening

**Comment by clinique** [comment]
Thanks because I’m really stuck 

Envoyé de mon iPhone

**Comment by J-N-K** [comment]
Automatic feature resolution is not working because we do not have the updated feature.xml at the moment. This will probably change in the next days. Installing the bundles manually will not work because we did not find out how to use wrap during manual install. 

The pom.xml is wrong anyway and I'll provide a PR for that.

File: features/karaf/openhab-addons/src/main/feature/feature.xml [hunk 3]
```diff
@@ -560,7 +560,13 @@
 
     <feature name="openhab-binding-netatmo" description="Netatmo Binding" version="${project.version}">
         <feature>openhab-runtime-base</feature>
-        <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.netatmo/${project.version}</bundle>
```
Comment (cweitkamp) [inline features/karaf/openhab-addons/src/main/feature/feature.xml]: @J-N-K Do we have to add the reference to wrap feature here too?

```
        <feature prerequisiste="true">wrap</feature>
```

**Comment by J-N-K** [comment]
Yes. I always forget that because we talked about adding it to openhab-runtime-base in the beginning.

File: features/karaf/openhab-addons/src/main/feature/feature.xml (hunk 3 above)
Comment (clinique) [reply features/karaf/openhab-addons/src/main/feature/feature.xml]: Should I submit a PR for this ?

**Comment by clinique** [comment]
> Automatic feature resolution is not working because we do not have the updated feature.xml at the moment. This will probably change in the next days. Installing the bundles manually will not work because we did not find out how to use wrap during manual install.
> 
> The pom.xml is wrong anyway and I'll provide a PR for that.

Having installed the snapshot version of OH2.5 this morning, I can confirm that this migration now works fine.

**Pull request merged**